File: blog_app/src/main/java/com/example/blog_app/service/BlogService.java
package com.example.blog_app.service;

import com.example.blog_app.model.Blog;
import com.example.blog_app.model.User;
import com.example.blog_app.model.dto.BlogDto;

import java.util.List;

public interface BlogService {
    Blog addBlog(BlogDto blogDto, User user);

    void deleteBlog(Long id);

    Blog updateBlog(Long id, BlogDto blogDto);

    List<Blog> getAllBlogs();

    Blog getBlogById(Long id);

    List<Blog> getBlogsByCategoryId(Long categoryId);

    List<Blog> getBlogsByAuthor(User author);

    List<Blog> getBlogsByAuthorAndCategory(User author, Long categoryId);

    Blog likeBlog(Long id, User currentUser);

    Blog dislikeBlog(Long id, User currentUser);

    boolean isLikedByUser(Long blogId, User currentUser);
}

File: frontend/src/api/services/BlogService.ts
import { Blog } from "../../types/Blogs";
import { BlogAppClientInstance } from "../rest-client";

export class BlogServices {
  getAllBlogs = async () => {
    return await BlogAppClientInstance.get<Blog[]>(`blogs`);
  };

  addBlog = async (formData: FormData) => {
    return await BlogAppClientInstance.post<Blog>(`blogs/add`, formData, {
      Accept: "*/*",
    });
  };

  updateBlog = async (id: string, formData: FormData) => {
    return await BlogAppClientInstance.put<Blog>(
      `blogs/update/${id}`,
      formData,
      {
        Accept: "*/*",
      }
    );
  };

  deleteBlog = async (id: string) => {
    return await BlogAppClientInstance.delete<void>(`blogs/delete/${id}`);
  };

  getBlogById = async (id: string) => {
    return await BlogAppClientInstance.get<Blog>(`blogs/${id}`);
  };

  getBlogsByCategory = async (categoryId: string) => {
    return await BlogAppClientInstance.get<Blog[]>(
      `blogs/category/${categoryId}`
    );
  };

  getMyBlogs = async () => {
    return await BlogAppClientInstance.get<Blog[]>(`blogs/myBlogs`);
  };

  getMyBlogsByCategory = async (categoryId: string) => {
    return await BlogAppClientInstance.get<Blog[]>(
      `blogs/myBlogs/${categoryId}`
    );
  };

  likeBlog = async (id: string) => {
    return await BlogAppClientInstance.post<Blog>(`blogs/like/${id}`, {});
  };

  dislikeBlog = async (id: string) => {
    return await BlogAppClientInstance.post<Blog>(`blogs/dislike/${id}`, {});
  };

  isLikedByUser = async (id: string) => {
    return await BlogAppClientInstance.get<boolean>(`blogs/${id}/liked`);
  };

  getLikedBlogsByUser = async () => {
    return await BlogAppClientInstance.get<Blog[]>(`blogs/likedByUser`);
  };
}

export const BlogService = new BlogServices();
